File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        p=q=head
        for i in range(n):
            p=p.next

        if q.next==p and p==None:
            return None
        elif p==None and q==head:
            head=head.next
        else:
            while(p.next!=None):
                p=p.next
                q=q.next
            q.next=q.next.next
        return head

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if(head == null || head.next == null) return null;
        ListNode p=head, q=head;
        for(int i = 0; i< n; i++){
            p = p.next;
        }
        if(p == null && q== head){
            return head.next;
        }

        while( p != null && p.next != null){
            p = p.next;
            q = q.next;
        }
        q.next= q.next.next;
        return head;
    }
}